pkg/image/resolve.go

```diff
@@ -188,3 +132,126 @@ func walkthroughMap(inputMap map[interface{}]interface{}, walkFunc func(map[inte
 		}
 	}
 }
+
+func GetImages(systemChartPath string, imagesFromArgs []string, rkeSystemImages map[string]v3.RKESystemImages, osType OSType) ([]string, error) {
+	var images []string
+
+	// fetch images from charts
+	if systemChartPath != "" {
+		imagesInCharts, err := fetchImagesFromCharts(systemChartPath, osType)
+		if err != nil {
+			return []string{}, errors.Wrap(err, "failed to fetch images from charts")
+		}
+		images = append(images, imagesInCharts...)
+	}
+
+	// fetch images from system images
+	if len(rkeSystemImages) > 0 {
+		imagesInSystem, err := fetchImagesFromSystem(rkeSystemImages, osType)
+		if err != nil {
+			return []string{}, errors.Wrap(err, "failed to fetch images from system images")
+		}
+		images = append(images, imagesInSystem...)
+	}
+
+	// append images from requirement
+	images = appendImagesFromRequirement(images, osType)
+
+	// append images from args
+	if len(imagesFromArgs) > 0 {
+		images = append(images, imagesFromArgs...)
+	}
+
+	return normalizeImages(images), nil
+}
+
+func fetchImagesFromCharts(path string, osType OSType) ([]string, error) {
+	chartVersion, err := getChartAndVersion(path)
+	if err != nil {
+		return []string{}, errors.Wrapf(err, "failed to get chart and version from %q", path)
+	}
+
+	imagesSet := map[string]struct{}{}
+	err = filepath.Walk(path, func(p string, info os.FileInfo, err error) error {
+		return pickImagesFromValuesYAML(imagesSet, chartVersion, path, p, info, err, osType)
+	})
+	if err != nil {
+		return []string{}, errors.Wrap(err, "failed to pick images from values.yaml")
+	}
+
+	var images []string
+	for image := range imagesSet {
+		images = append(images, image)
+	}
+	return images, nil
+}
+
+func fetchImagesFromSystem(rkeSystemImages map[string]v3.RKESystemImages, osType OSType) ([]string, error) {
+	collectionImagesList := []interface{}{
+		rkeSystemImages,
+	}
+	switch osType {
+	case Linux:
+		collectionImagesList = append(collectionImagesList, v3.ToolsSystemImages)
+	}
+
+	return flatImagesFromCollections(collectionImagesList...)
+}
+
+func flatImagesFromCollections(cols ...interface{}) (images []string, err error) {
+	for _, col := range cols {
+		colObj := map[string]interface{}{}
+		if err := convert.ToObj(col, &colObj); err != nil {
+			return []string{}, err
+		}
+
+		images = append(images, fetchImagesFromCollection(colObj)...)
+	}
+	return images, nil
+}
+
+func fetchImagesFromCollection(obj map[string]interface{}) (images []string) {
+	for _, v := range obj {
+		switch t := v.(type) {
+		case string:
+			images = append(images, t)
+		case map[string]interface{}:
+			images = append(images, fetchImagesFromCollection(t)...)
+		}
+	}
+	return images
+}
+
+func appendImagesFromRequirement(images []string, osType OSType) []string {
+	switch osType {
+	case Linux:
+		images = append(images, linuxRequiredImagesNotInSystemCharts...)
+	}
+	return images
+}
+
+func normalizeImages(rawImages []string) (images []string) {
```

Comment:
nit: maybe we don't need named return value here since we're as it is declaring uImages and imagesSet within this method?

Reply:
Updated